### PR TITLE
Improve StationJobsSystem.AssignJobs readability

### PR DESCRIPTION
**Pull request by Absotively**
## About the PR
I refactored `StationJobsSystem.AssignJobs` with a focus on readability. I also added more tests for it, and tweaked the existing tests.

I am fairly confident that I have avoided actually changing the job assignment logic at all.

## Why / Balance
I have seen multiple people, in multiple discords, assert that job assignment logic is too complex to try to understand, or too complex to risk changing. Eventually I got fed up with this function scaring people off.

The actual job assignment logic is not that complex, but this function was overtuned in a way that made it seem more complex. It also had a comment discouraging people from attempting to understand it, which I believe contributed to the impression of complexity. I hope that refactoring it for readability will make contributors less hesitant to consider improvements to job assignment.

The previous tests for job assignment left many edge cases uncovered, and I wanted more thorough tests to confirm this refactor wasn't breaking things. However, the job assignment logic doesn't actually handle all its edge cases well, especially when they involve multiple stations. Rather than writing tests for behaviour that should probably be changed, I only added enough tests to achieve good coverage of the old version of this function according to measurements from `dotnet-coverage`.

## Technical details
* Renamed many variables
* Rewrote many comments
* Removed redundant bookkeeping
* Reworked the AssignPlayer helper function
    * Inlined most of it, so that it's easier to see how the bookkeeping is being updated
    * Renamed the function that still does the remaining slightly-more-complex bookkeeping task to `RemoveJobCandidate`
    * Moved the function definition out of the AssignJobs function body
* Added a helper function, `CalculateStationShares`, to calculate how many players each station can claim in a given iteration. This code is a bit lengthy and very self-contained, so moving it out of the main function helps readability of both this and the main function, in my opinion.
* Moved the creation of a list of job slots with the current weight outside the loop over job priorities, since job priority doesn't change what jobs have the current weight
* Removed code for re-using collections to avoid allocations; the collections in question are now created fresh inside the relevant loops
* Converted loops to LINQ expressions in cases where the LINQ expression seemed clearer
* Modified flow control slightly to avoid `goto` statements and do-while loop

AssignJobsTest reports on the time taken by AssignJobs, and previously used a very large data set, as a sort of makeshift benchmark. I was not ambitious enough to turn this into a proper benchmark. I reduced the size of the dataset when I added more tests because using the same dataset for more tests caused problems with the test pair (cargo ran out of bounty IDs). I did use the reported time to check that my changes were not making this function substantially slower, and I omitted one refactor because it made the function slower.

So far as the speed goes, please note that this function takes on the order of tens of milliseconds, and is called exactly once per round, during round start, when everyone is waiting for map init anyways.

## Requirements

- [X] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [X] I have added media to this PR or it does not require an in-game showcase.

## Breaking changes

**Changelog**

### CODE REVIEW

**Comment by Absotively** [comment]
Test fail appears unrelated:
```
   Error Message:
   Multiple failures or warnings in test:
  1)   Assert.That(a.SkinColor, Is.EqualTo(b.SkinColor))
  Expected: {(R, G, B, A) = (0.5686275, 0.30980393, 0.14117648, 1)}
  But was:  {(R, G, B, A) = (0.5647059, 0.30980393, 0.14117648, 1)}
```

File: Content.Server/Station/Systems/StationJobsSystem.Roundstart.cs [hunk 1]
```diff
@@ -47,12 +47,10 @@ private void InitializeRoundStart()
     /// </summary>
     /// <param name="profiles">The profiles to use for selection.</param>
     /// <param name="stations">List of stations to assign for.</param>
-    /// <param name="useRoundStartJobs">Whether or not to use the round-start jobs for the stations instead of their current jobs.</param>
+    /// <param name="useRoundStartJobs">Whether or not to use the round-start jobs for the stations instead of their
```
Comment (moonheart08) [inline Content.Server/Station/Systems/StationJobsSystem.Roundstart.cs:50]: This is not as good an explainer as the old one. You can use round-start jobs mid-round just fine, you simply shouldn't for *already set up and in use* stations. If you for example create a new station with a pool of players midround in your fork, you should set this.

File: Content.Server/Station/Systems/StationJobsSystem.Roundstart.cs [hunk 2]
```diff
@@ -63,213 +61,196 @@ private void InitializeRoundStart()
...
+
+                    // We want to go through them in random order.
+                    _random.Shuffle(currentJobs);
+
+                    // Loop through the jobs repeatedly until one of the following happens:
+                    // * The station has its share of players for the current weight & priority
+                    // * All the players who were in jobCandidates have been assigned jobs
+                    // * None of the remaining jobCandidates can be assigned to any of the jobs in currentJobs, due
+                    //   to the jobs being full and/or the players not matching the remaining jobs
+                    var stillAssigningJobs = true;
```
Comment (moonheart08) [inline Content.Server/Station/Systems/StationJobsSystem.Roundstart.cs:142]: This isn't more readable than the goto. Even if people seem to dislike goto, that doesn't change that it's the right control flow choice here and this fancy bool is doing the same thing but worse, with more fallthrough required to understand what flow it is replicating.

File: Content.Server/Station/Systems/StationJobsSystem.Roundstart.cs [hunk 3]
```diff
@@ -63,213 +61,196 @@ private void InitializeRoundStart()
...
-                    // Oh the misery, another allocation.
-                    var allJobs = currStationSelectingJobs.Keys.ToList();
-                    _random.Shuffle(allJobs);
-                    // And iterates through all it's jobs in a random order until the count settles.
-                    // No, AFAIK it cannot be done any saner than this. I hate "shaking" collections as much
-                    // as you do but it's what seems to be the absolute best option here.
-                    // It doesn't seem to show up on the chart, perf-wise, anyway, so it's likely fine.
-                    int priorCount;
-                    do
+                    var currentJobs = currentWeightJobSlots[station].Keys.ToList();
```
Comment (moonheart08) [inline Content.Server/Station/Systems/StationJobsSystem.Roundstart.cs]: Nit: Caching this index got removed, and while it's not critical, it's now a bunch of repetition over just having currStationSelectingJobs.

File: Content.Server/Station/Systems/StationJobsSystem.Roundstart.cs [hunk 4]
```diff
@@ -63,213 +61,196 @@ private void InitializeRoundStart()
...
         }
 
-        endFunc:
         return assigned;
     }
 
+    private static void RemoveJobCandidate(Dictionary<ProtoId<JobPrototype>, HashSet<NetUserId>> jobCandidates,
+        NetUserId candidateToRemove)
+    {
+        // Remove the player from all possible jobs as that's faster than actually checking what they have selected.
```
Comment (moonheart08) [inline Content.Server/Station/Systems/StationJobsSystem.Roundstart.cs]: comment is wrong, it's not faster just simpler. Yes this is my very own fault.

```suggestion
        // Remove the player from all possible jobs as that's simpler than actually checking what they have selected and inconsequential.
```

**Review by moonheart08 — CHANGES_REQUESTED** [review]
Algorithm changes look mostly fine, I'm a little curious about why you inlined assignplayer but then outlined RemoveJobCandidate and CalculateStationShares. (I think all three should be their own functions, not just two of 'em)

I do appreciate the additional tests.

**Comment by Absotively** [comment]
> Algorithm changes look mostly fine, I'm a little curious about why you inlined assignplayer but then outlined RemoveJobCandidate and CalculateStationShares. (I think all three should be their own functions, not just two of 'em)

I initially just moved the definition of AssignPlayer out of AssignJobs, but then of course I had to pass in all the bookkeeping variables as parameters, which made the function call long enough that it made sense to put it on multiple lines. At that point, I figured that if the inner loop was going to include a line for every bookkeeping variable, it might as well have the bookkeeping on that line.

I also think it's useful to have the simpler bookkeeping updates near the checks that use many of the bookkeeping variables, so it's easy to see how the two correlate.

**Comment by moonheart08** [comment]
I need to make some changes to the algorithm myself so I'm just going to fork off this PR into a new PR.

**Comment by Absotively** [comment]
Ah, I missed the last comment! I am entirely fine with this being superseded by #43002.

I had a thought for how to make the last loop actually clearer, but it made the test a small but non-negligable amount slower, so then I got bogged down trying to figure out things involving dotnet-trace. I will try to get it cleaned up a bit and committed tonight, in case it still proves useful, but if it goes unused because I was slow, that is fine.

File: Content.Server/Station/Systems/StationJobsSystem.Roundstart.cs (hunk 2 above)
Comment (Absotively) [reply Content.Server/Station/Systems/StationJobsSystem.Roundstart.cs:142]: So after going back and forth with various options, I decided that I do prefer `stillAssigningJobs` to `priorCount` but combining it with the inner loop break/goto conditions was confusing.

Using `break` instead of `goto` at the top of the inner loop is in theory a little slower without the combined condition, but only a little, and I do think it's slightly easier to reason about because it doesn't require the reader to look for and remember where control is going. So I have left that.

Upon review, another problem was that I was overexplaining things that are actually pretty straightforward, or not relevant at the point of the explanation, so I've shortened some of the comments here.

**Comment by Absotively** [comment]
This is now ready for review again, if the maints want to review it and #43002's algorithm changes separately. If the maints instead prefer to close this and review the two sets of changes together in #43002, that is also fine with me.
